File: examples/single_trial_ari.py
from __future__ import annotations

from statistics import fmean

import raven

import _helpers as h


# Raven parameters:
CORESET_SIZE = 256
SAMPLING_SEEDS = 64
NUM_TRIALS = 1
RAVEN_RNG_SEED = 42


# Workload parameters:
WORKLOAD_SEED = 42


workload = h.make_workload(seed=WORKLOAD_SEED)

# initialize Raven index:
index = raven.Raven(
    h.NUM_CLUSTERS,
    coreset_size=CORESET_SIZE,
    sampling_seeds=SAMPLING_SEEDS,
    num_trials=NUM_TRIALS,
    rng_seed=RAVEN_RNG_SEED,
    **h.INDEX_KWARGS,
)

scores: list[float] = []

for batch_index, batch in enumerate(workload.batches):

    # edge_ops is a list of (u, v, weight) tuples, where weight=None indicates deletion
    index.update_edges(batch.edge_ops)

    # flush the index to apply all updates before querying
    # (triggers on query anyway)
    index.flush()

    # get info about the current snapshot
    # including which nodes the workload wants to query + their true labels
    snapshot = h.query_snapshot(
        index,
        workload,
        batch_index,
        batch,
        coreset_size=CORESET_SIZE,
        workload_seed=WORKLOAD_SEED,
    )
    if snapshot is None:
        continue

    # query the index
    result = index.query(snapshot.query_nodes)
    score = h.ari(snapshot.true_labels, result.labels)
    scores.append(score)
    print(
        f"batch={snapshot.batch_index:02d} "
        f"time={snapshot.batch_time} "
        f"updates={len(batch.edge_ops)} "
        f"nodes={len(snapshot.query_nodes)} "
        f"ari={score:.3f}"
    )

print(f"mean ARI: {fmean(scores):.3f}")

File: crates/raven/examples/single_trial_ari.rs
use raven::{EdgeUpdate, Raven, RavenConfig};
use raven_core::metrics::adjusted_rand_index;

fn main() -> Result<(), raven::RavenError> {
    let mut config = RavenConfig::new(2);
    config.coreset_size = 4;
    config.sampling_seeds = 2;
    config.num_trials = 1;
    config.rng_seed = Some(42);
    config.node_capacity = 16;
    config.expected_edges_per_node = 4;

    let mut index = Raven::new(config)?;
    let query_nodes = [1, 2, 3, 10, 11, 12];
    let true_labels = [0, 0, 0, 1, 1, 1];

    let batches = [
        vec![
            EdgeUpdate::set(1, 2, 1.0),
            EdgeUpdate::set(2, 3, 1.0),
            EdgeUpdate::set(10, 11, 1.0),
            EdgeUpdate::set(11, 12, 1.0),
        ],
        vec![EdgeUpdate::set(1, 3, 1.0), EdgeUpdate::set(10, 12, 1.0)],
    ];

    for (batch_index, batch) in batches.into_iter().enumerate() {
        index.update_edges(batch)?;
        index.flush()?;

        let result = index.query(&query_nodes)?;
        let ari = adjusted_rand_index(&true_labels, &result.labels);
        println!(
            "batch={batch_index} labels={:?} ari={ari:.3}",
            result.labels
        );
    }

    Ok(())
}
